source/common/filter/tcp_proxy.h

```diff
@@ -133,6 +138,46 @@ class TcpProxy : public Network::ReadFilter, Logger::Loggable<Logger::Id::filter
     bool on_high_watermark_called_{false};
   };
 
+  struct UpstreamHelperCallbacks {
+    virtual ~UpstreamHelperCallbacks() {}
+
+    virtual const std::string& getUpstreamCluster() PURE;
```

Comment:
Is there a reason that we need to have a new object here that we allocate? Can these just be normal virtual functions that get overriden by the WsHandlerImpl derived class? That would save the extra allocation which AFAICT isn't needed?